### DISCUSSION TITLE
Where to get/how to generate config instance (to use with tauri::api::path::app_data_dir)?

### DISCUSSION THREAD

**Discussion by StelFoog** [tauri-apps/tauri#6583] [Q&A] [answered]
I'm new to Tauri so this is probably a very obvious question, but I'm trying to get started by creating a file in the `app_data_dir` (which I understand to be the Tauri equivalent of Electron's `app.getPath('userData')`). I'm trying to get the `app_data_dir` path using `tauri::api::path::app_data_dir(&config)`, but I don't know what value I should use for `config`. From what I read I interpret this as a struct which represents `tauri.config.json` so I should read it from somewhere instead of creating a new instance of it. The only way I found to get it was using an old [`tauri_api::config.get()`](https://docs.rs/tauri-api/latest/tauri_api/config/fn.get.html) which seems to be deprecated. What's the correct way to get the config instance?

**Comment by FabianLars** [comment, marked as answer]
The idea is to get it from an instance of `App` or `AppHandle` which you can get in the `setup` hook and [commands](https://tauri.app/v1/guides/features/command#accessing-an-apphandle-in-commands) for example.
And if you have an app instance there's a helper function to make it a bit more streamlined (without the config stuff): `app.path_resolver().app_data_dir()`

**Comment by aleksey-hoffman** [comment]
Here's the proper way to get all the paths relevant to your app:
Docs are not helpful. Spent like an hour figuring it out for my app

You have to use 2 different sources to get all the paths:

```rust
fn main() {
    tauri::Builder::default()
        .setup(setup_handler)
        .invoke_handler(tauri::generate_handler![])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}

fn setup_handler(app: &mut tauri::App) -> Result<(), Box<dyn std::error::Error + 'static>> {
    let app_handle = app.handle();

    println!("{}", app_handle.path_resolver().resource_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", app_handle.path_resolver().app_config_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", app_handle.path_resolver().app_data_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", app_handle.path_resolver().app_local_data_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", app_handle.path_resolver().app_cache_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", app_handle.path_resolver().app_log_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::data_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::local_data_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::cache_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::config_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::executable_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::public_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::runtime_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::template_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::font_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::home_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::audio_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::desktop_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::document_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::download_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
    println!("{}", tauri::api::path::picture_dir().unwrap_or(std::path::PathBuf::new()).to_string_lossy());
}
```

**Reply by natestraub** [reply]
Honestly thanks a million, the docs are so bad for this.

**Reply by dukedorje** [reply]
Wow, seriously, thank you so much. Tauri is so awesome but SO HARD TO FIGURE OUT how to use it -- if you are on a timeline, you're gonna drop Tauri & just go with some crap like Electron. How can we improve this situation?

**Reply by aleksey-hoffman** [reply]
@dukejones I feel you man, I'm rewriting my large desktop app ([Sigma File Manager](https://github.com/aleksey-hoffman/sigma-file-manager)) moving from Electron to Tauri, and it's unfortunate that the lack of good documentation is slowing down the process. A lot of people want to see it, but I cannot migrate it quickly enough by myself

**Reply by jamesfebin** [reply]
Thank you, I was facing the same issue, docs are not helpful.

**Reply by karimelhajoui63** [reply]
I don't know much about Rust, but for me, the `Ok(())` was missing at the very end of the `setup_handler` function

**Comment by Ledich19** [comment]
this works for me
the documentation is really terrible))
```rs
pub fn run() {
    tauri::Builder::default()
        .setup(|app| {
            let path = app.path().data_dir();
            println!("----{:?}-----", path.unwrap());
            Ok(())
        })
        .plugin(tauri_plugin_fs::init())
        .plugin(tauri_plugin_shell::init())
        .invoke_handler(tauri::generate_handler![
            greet,
            commands::commands::my_custom_command,
            commands::commands::my_custom_command_2
        ])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

**Reply by FabianLars** [reply]
>  the documentation is really terrible))

contributions are always welcome :))

(srsly, so many complaints + examples here and nobody wants to fix it for others?)

**Reply by dukedorje** [reply]
this is a good point. i've also complained in the past. if you find a way, help to clear the path for the next person. that's the beauty of open source. i'll be contributing to docs as i do more with tauri.
